<a href="https://colab.research.google.com/github/ananyapoonsawat/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/BSC_DPDM23/'

Mounted at /content/drive


In [3]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM_23/BSC_dpdm23_data/reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [ ]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

# แปลง Data ให้อยู่ในรูปแบบ Transaction data ด้วย $groupby.apply(list)$

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list) #เอาตัวแปรมาเก็บ

เตรียม data เสร็จแล้ว

# ทำ Association Rules

In [ ]:
!pip install apyori #ต้อง install แพ๊กเกตใหม่ เพราะไม่ใช่ทั่วๆไปที่คนใช้ install ครั้งเดียวพอ

In [ ]:
from apyori import apriori

In [ ]:
apriori(transactions,min_support = 0.065)

<generator object apriori at 0x7858b41dd1c0>

In [ ]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
frequentItemsets[0][0]

frozenset({'22423'})

In [ ]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
frequentItemsets[0][0].__len__()

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

## Quiz 22386 และ 85099B คือสินค้าอะไร

In [ ]:
set(data[data['StockCode']=='22386' ]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [ ]:
set(data[data['StockCode']=='85099B' ]['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

## HW 1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ+เดือน

สั่งวันที 3 มกราคม 2567
ส่งก่อน วันจันทร์หน้า

##เตรียมข้อมูล

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

เราต้องการแค่ เดือน/วัน/ปี ใน Column InvoiceData ไม่ได้ต้องการ เวลา

แยก เดือน/วัน/ปี ออกจาก เวลา ด้วยการทำ Prepare Data

## ทำ Data Preparation

Data Preparation คือ การแปลง(Transform)ข้อมูลให้อยู่ในรูปแบบที่เหมาะสม


In [ ]:
data['InvoiceDate'][0].split()[0]

'12/01/2010'

In [ ]:
date = []
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)
#set(L)

In [ ]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

In [ ]:
# Test
date[0].split('/')

['12', '01', '2010']

## นำ Data ที่ผ่านการแปลงแล้วไปใช้

>



In [ ]:
data_test_nonclean = data

In [ ]:
data_test_nonclean['date'] = date
data_test_nonclean['day'] = day
data_test_nonclean['month'] = month
data_test_nonclean['year'] = year

# data_test_nonclean >>> ข้อมูลยังไม่อยู่ในมาตรฐานเดียวกัน

In [ ]:
set(data_test_nonclean['date'])

{'01/04/2011',
 '01/05/2011',
 '01/06/2011',
 '01/07/2011',
 '01/09/2011',
 '01/10/2011',
 '01/11/2011',
 '01/12/2011',
 '02/01/2011',
 '02/02/2011',
 '02/03/2011',
 '02/04/2011',
 '02/06/2011',
 '02/07/2011',
 '02/08/2011',
 '02/09/2011',
 '02/10/2011',
 '02/11/2011',
 '03/01/2011',
 '03/02/2011',
 '03/03/2011',
 '03/04/2011',
 '03/06/2011',
 '03/07/2011',
 '03/08/2011',
 '03/09/2011',
 '03/10/2011',
 '03/11/2011',
 '04/01/2011',
 '04/03/2011',
 '04/04/2011',
 '04/05/2011',
 '04/06/2011',
 '04/07/2011',
 '04/08/2011',
 '04/10/2011',
 '04/11/2011',
 '04/12/2011',
 '05/01/2011',
 '05/03/2011',
 '05/04/2011',
 '05/05/2011',
 '05/06/2011',
 '05/08/2011',
 '05/09/2011',
 '05/10/2011',
 '05/11/2011',
 '05/12/2011',
 '06/01/2011',
 '06/02/2011',
 '06/03/2011',
 '06/05/2011',
 '06/06/2011',
 '06/07/2011',
 '06/08/2011',
 '06/09/2011',
 '06/10/2011',
 '06/12/2011',
 '07/01/2011',
 '07/03/2011',
 '07/04/2011',
 '07/05/2011',
 '07/06/2011',
 '07/07/2011',
 '07/08/2011',
 '07/10/2011',
 '07/11/20

In [ ]:
set(data_test_nonclean['day'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

### แปลงวัน ให้เป็นมาตรฐานเดียวกัน

วัน เป็นมาตรฐานเดียวกันไม่ต้องแปลง

In [ ]:
data_test_nonclean.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010  01   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010  01   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

### แปลงเดือน ให้เป็นมาตรฐานเดียวกัน

In [ ]:
set(data_test_nonclean['month'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '1',
 '10',
 '11',
 '12',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9'}

In [ ]:
data_test = data_test_nonclean.replace(to_replace=['1', '2','3','4','5','6','7','8','9','10','11','12'],value=[1,2,3,4,5,6,7,8,9,10,11,12])

In [ ]:
data_test.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
date           False
day            False
month          False
year           False
dtype: bool

In [ ]:
data_test_nonclean[data_test_nonclean['month']=='01']

InvoiceNo StockCode                Description  Quantity  \
42481    539993     22386    JUMBO BAG PINK POLKADOT        10   
42482    539993     21499         BLUE POLKADOT WRAP        25   
42483    539993     21498        RED RETROSPOT WRAP         25   
42484    539993     22379   RECYCLING BAG RETROSPOT          5   
42485    539993     20718  RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                        ...       ...   
55385    540992     21912   VINTAGE SNAKES & LADDERS         2   
55386    540992     51008   AFGHAN SLIPPER SOCK PAIR        10   
55387    540992     21508    VINTAGE KID DOLLY CARD         12   
55388    540992     22716         CARD CIRCUS PARADE        12   
55389    540992     22983        CARD BILLBOARD FONT        12   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
55385  01/12/2011 16:57       3.75     17364.0  United Kingdom  01/12/2011   
55386  01/12/2011 16:57       3.45     17364.0  United Kingdom  01/12/2011   
55387  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55388  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55389  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   

      day month  year  
42481  04    01  2011  
42482  04    01  2011  
42483  04    01  2011  
42484  04    01  2011  
42485  04    01  2011  
...    ..   ...   ...  
55385  12    01  2011  
55386  12    01  2011  
55387  12    01  2011  
55388  12    01  2011  
55389  12    01  2011  

[12909 rows x 12 columns]

In [ ]:
data_test[data_test['month']==1]

InvoiceNo StockCode                      Description  Quantity  \
55390    540993     22423         REGENCY CAKESTAND 3 TIER         2   
55391    540993     22776      SWEETHEART CAKESTAND 3 TIER         1   
55392    540993     21471     STRAWBERRY RAFFIA FOOD COVER         6   
55393    540993     47580            TEA TIME DES TEA COSY         6   
55394    540993     21212  PACK OF 72 RETROSPOT CAKE CASES        24   
...         ...       ...                              ...       ...   
77623    542745     22138    BAKING SET 9 PIECE RETROSPOT         13   
77624    542746     21034     REX CASH+CARRY JUMBO SHOPPER         2   
77625   C542748     22969     HOMEMADE JAM SCENTED CANDLES       -12   
77626   C542753     22649     STRAWBERRY FAIRY CAKE TEAPOT        -1   
77627    542759     22649     STRAWBERRY FAIRY CAKE TEAPOT         1   

           InvoiceDate  UnitPrice  CustomerID         Country       date day  \
55390   1/13/2011 9:23      12.75     13284.0  United Kingdom  1/13/2011  13   
55391   1/13/2011 9:23       9.95     13284.0  United Kingdom  1/13/2011  13   
55392   1/13/2011 9:23       3.75     13284.0  United Kingdom  1/13/2011  13   
55393   1/13/2011 9:23       2.55     13284.0  United Kingdom  1/13/2011  13   
55394   1/13/2011 9:23       0.55     13284.0  United Kingdom  1/13/2011  13   
...                ...        ...         ...             ...        ...  ..   
77623  1/31/2011 16:47       4.95     15881.0  United Kingdom  1/31/2011  31   
77624  1/31/2011 16:48       0.95     15881.0  United Kingdom  1/31/2011  31   
77625  1/31/2011 16:53       1.45     15881.0  United Kingdom  1/31/2011  31   
77626  1/31/2011 16:57       4.95     15881.0  United Kingdom  1/31/2011  31   
77627  1/31/2011 16:59       4.95     15881.0  United Kingdom  1/31/2011  31   

      month  year  
55390     1  2011  
55391     1  2011  
55392     1  2011  
55393     1  2011  
55394     1  2011  
...     ...   ...  
77623     1  2011  
77624     1  2011  
77625     1  2011  
77626     1  2011  
77627     1  2011  

[22238 rows x 12 columns]

###แยกประเทศออกตาม ทวีป

In [ ]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

#### ทวีปยุโรป

In [ ]:
EU_country = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

#### ทวีปเอเชีย

In [ ]:
Asia_country = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

## Code หาชื่อสินค้า จาก Column "Description"

In [ ]:
def finditem(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L

## 1. หา Frequent Item Set และ Association Rules ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)

### จัดกลุ่มตาม Transaction (Invoice_No)

In [ ]:
data_CM_and_NY = data_test[data_test['month']==12]

In [ ]:
data_CM_and_NY

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
541904  09    12  2011  
541905  09    12  2011  
541906  09    12  2011  
541907  09    12  2011  
541908  09    12  2011  

[68006 rows x 12 columns]

In [ ]:
transactionsCM_and_NY = data_CM_and_NY.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
frequentItemsets = list(apriori(transactionsCM_and_NY,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'22111'}), support=0.07203947368421053, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22111'}), confidence=0.07203947368421053, lift=1.0)]),
 RelationRecord(items=frozenset({'22112'}), support=0.07072368421052631, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22112'}), confidence=0.07072368421052631, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),
 RelationRecord(items=frozenset({'22910'}), support=0.07171052631578947, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [ ]:
len(frequentItemsets)

6

In [ ]:
min_sup = 0.035

frequentItemsets = list(apriori(transactionsCM_and_NY,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])
yes
RelationRecord(items=frozenset({'22835', '22112'}), support=0.036513157894736845, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22112', '22835'}), confidence=0.036513157894736845, lift=1.0), OrderedStatistic(items_base=frozenset({'22112'}), items_add=frozenset({'22835'}), confidence=0.5162790697674419, lift=9.748374981944245), OrderedStatistic(items_base=frozenset({'22835'}), items_add=frozenset({'22112'}), confidence=0.68944099378882, lift=9.748

In [ ]:
set(data[(data['StockCode']=='22086') | (data['StockCode']=='22910') ]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

In [ ]:
finditem('22086','22910')

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

มี 3 itemset ที่ผ่าน 0.035

RelationRecord(items=frozenset({'22086', '22910'}),
support=0.047697368421052634,
<br>
<br>
**คนที่ซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) กับ  22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 4.77% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) แล้วค่อยซื้อ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) มีค่า confidence เท่ากับ 51.6% มีค่า lift เท่ากับ 7.19 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331),
<br>
<br>
**คนที่ซื้อ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) แล้วค่อยซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) มีค่า confidence ตั้งเท่ากับ 66.51% มีค่า lift เท่ากับ 7.19 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])

เสร็จข้อ 1

## 2. หา Frequent Item Set และ Association Rules ของประเทศในทวีป Europe


### จัดกลุ่มประเทศใส่ในตัวแปรทวีป Europe

In [ ]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

In [ ]:
Eu = ['Austria','Belgium','Channel Islands','Cyprus','Czech Republic','Denmark','EIRE','European Community',
      'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
      'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

### จัดกลุ่มตาม Transaction (Invoice_No)

In [ ]:
Eu_data = data_test[data_test['Country']==""]
for i in range(len(Eu)):
  x = data_test[data_test['Country']==Eu[i]]
  Eu_data=pd.concat([x,Eu_data])

In [ ]:
Eu_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
536964  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536965  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536966  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536967  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536968  12/08/2011 10:26      40.00     12358.0         Austria  12/08/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
536964  08    12  2011  
536965  08    12  2011  
536966  08    12  2011  
536967  08    12  2011  
536968  08    12  2011  

[538358 rows x 12 columns]

In [ ]:
transactions_Eu = Eu_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
frequentItemsets = list(apriori(transactions_Eu,min_support = 0.085))
frequentItemsets

[RelationRecord(items=frozenset({'85123A'}), support=0.08713225292448797, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08713225292448797, lift=1.0)])]

In [ ]:
len(frequentItemsets)

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions_Eu,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03011931133652015, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03011931133652015, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7402101241642789, lift=17.20537191045263), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7000903342366757, lift=17

In [ ]:
set(data[(data['StockCode']=='85099B') | (data['StockCode']=='22386')]['Description'])

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

In [ ]:
finditem('85099B', '22386')

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

มี 2 itemset ที่ผ่าน 0.03

RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679,
<br>
<br>
**คนที่ซื้อ 85099B(JUMBO BAG PINK POLKADOT) กับ 22386(JUMBO BAG RED RETROSPOT) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 3.23% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22386(JUMBO BAG RED RETROSPOT) แล้วค่อยซื้อ 85099B(JUMBO BAG PINK POLKADOT) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 67.75% มีค่า lift เท่ากับ 8.20 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402),
<br>
<br>
**คนที่ซื้อ 85099B(JUMBO BAG PINK POLKADOT) แล้วค่อยซื้อ 22386(JUMBO BAG RED RETROSPOT)  จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 39.15% มีค่า lift เท่ากับ 8.20 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes

ข้อ 2 เสร็จแล้ว

## 3.หา Frequent Item Set และ Association Rules ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)

### จัดกลุ่มประเทศใส่ในตัวแปรทวัป Asia

In [ ]:
Asia= ['Bahrain','Hong Kong','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [ ]:
Asia_data = data_test[data_test['Country']==""]
for i in range(len(Asia)):
  x = data_test[data_test['Country']==Asia[i]]
  Asia_data=pd.concat([x,Asia_data])

In [ ]:
Asia_data

InvoiceNo StockCode                         Description  Quantity  \
89570     543911     21485    RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                  DOORMAT UNION FLAG         2   
89572     543911     85150       LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                 WASHROOM METAL SIGN        24   
89574     543911     82551              LAUNDRY 15C METAL SIGN        12   
...          ...       ...                                 ...       ...   
181152    552449    72802C   VANILLA SCENT CANDLE JEWELLED BOX         6   
197371    553900     37449  CERAMIC CAKE STAND + HANGING CAKES         2   
197372    553900     37446  MINI CAKE STAND WITH HANGING CAKES         8   
197373    553900     22890  NOVELTY BISCUITS CAKE STAND 3 TIER         4   
197374    553900     37450   CERAMIC CAKE BOWL + HANGING CAKES         6   

             InvoiceDate  UnitPrice  CustomerID               Country  \
89570    2/14/2011 12:46       4.95     17829.0  United Arab Emirates   
89571    2/14/2011 12:46       7.95     17829.0  United Arab Emirates   
89572    2/14/2011 12:46       2.55     17829.0  United Arab Emirates   
89573    2/14/2011 12:46       1.45     17829.0  United Arab Emirates   
89574    2/14/2011 12:46       1.45     17829.0  United Arab Emirates   
...                  ...        ...         ...                   ...   
181152  05/09/2011 13:49       4.25     12355.0               Bahrain   
197371   5/19/2011 17:47       9.95     12353.0               Bahrain   
197372   5/19/2011 17:47       1.45     12353.0               Bahrain   
197373   5/19/2011 17:47       9.95     12353.0               Bahrain   
197374   5/19/2011 17:47       2.95     12353.0               Bahrain   

              date day month  year  
89570    2/14/2011  14     2  2011  
89571    2/14/2011  14     2  2011  
89572    2/14/2011  14     2  2011  
89573    2/14/2011  14     2  2011  
89574    2/14/2011  14     2  2011  
...            ...  ..   ...   ...  
181152  05/09/2011  09    05  2011  
197371   5/19/2011  19     5  2011  
197372   5/19/2011  19     5  2011  
197373   5/19/2011  19     5  2011  
197374   5/19/2011  19     5  2011  

[1314 rows x 12 columns]

### หาช่วงเวลาตรุษจีน

In [ ]:
Asia_data_Jan = Asia_data[(Asia_data['month']==1)]

In [ ]:
Asia_data_Jan

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   
...         ...       ...                               ...       ...   
69675    541975     20726                LUNCH BAG WOODLAND        20   
69676    541975     22452      MEASURING TAPE BABUSHKA PINK        24   
69677    541975     22454       MEASURING TAPE BABUSHKA RED        24   
69678    541975     20749        ASSORTED COLOUR MINI CASES         6   
69679    541975     20682  RED RETROSPOT CHILDRENS UMBRELLA         6   

           InvoiceDate  UnitPrice  CustomerID    Country       date day month  \
70758  1/25/2011 13:26       0.36     12744.0  Singapore  1/25/2011  25     1   
70759  1/25/2011 13:26       2.95     12744.0  Singapore  1/25/2011  25     1   
70760  1/25/2011 13:26       2.95     12744.0  Singapore  1/25/2011  25     1   
70761  1/25/2011 13:26       0.21     12744.0  Singapore  1/25/2011  25     1   
70762  1/25/2011 13:26       0.21     12744.0  Singapore  1/25/2011  25     1   
...                ...        ...         ...        ...        ...  ..   ...   
69675  1/24/2011 14:24       1.65         NaN  Hong Kong  1/24/2011  24     1   
69676  1/24/2011 14:24       2.95         NaN  Hong Kong  1/24/2011  24     1   
69677  1/24/2011 14:24       2.95         NaN  Hong Kong  1/24/2011  24     1   
69678  1/24/2011 14:24       7.95         NaN  Hong Kong  1/24/2011  24     1   
69679  1/24/2011 14:24       3.25         NaN  Hong Kong  1/24/2011  24     1   

       year  
70758  2011  
70759  2011  
70760  2011  
70761  2011  
70762  2011  
...     ...  
69675  2011  
69676  2011  
69677  2011  
69678  2011  
69679  2011  

[158 rows x 12 columns]

In [ ]:
Asia_data_Jan['day'] = pd.to_numeric(Asia_data_Jan['day'], errors='coerce')

<ipython-input-197-9bafe383ada4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Asia_data_Jan['day'] = pd.to_numeric(Asia_data_Jan['day'], errors='coerce')


In [ ]:
Asia_data_DJ = Asia_data_Jan[(Asia_data_Jan['day'] >= 15)]

In [ ]:
Asia_data_DJ

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   
...         ...       ...                               ...       ...   
69675    541975     20726                LUNCH BAG WOODLAND        20   
69676    541975     22452      MEASURING TAPE BABUSHKA PINK        24   
69677    541975     22454       MEASURING TAPE BABUSHKA RED        24   
69678    541975     20749        ASSORTED COLOUR MINI CASES         6   
69679    541975     20682  RED RETROSPOT CHILDRENS UMBRELLA         6   

           InvoiceDate  UnitPrice  CustomerID    Country       date  day  \
70758  1/25/2011 13:26       0.36     12744.0  Singapore  1/25/2011   25   
70759  1/25/2011 13:26       2.95     12744.0  Singapore  1/25/2011   25   
70760  1/25/2011 13:26       2.95     12744.0  Singapore  1/25/2011   25   
70761  1/25/2011 13:26       0.21     12744.0  Singapore  1/25/2011   25   
70762  1/25/2011 13:26       0.21     12744.0  Singapore  1/25/2011   25   
...                ...        ...         ...        ...        ...  ...   
69675  1/24/2011 14:24       1.65         NaN  Hong Kong  1/24/2011   24   
69676  1/24/2011 14:24       2.95         NaN  Hong Kong  1/24/2011   24   
69677  1/24/2011 14:24       2.95         NaN  Hong Kong  1/24/2011   24   
69678  1/24/2011 14:24       7.95         NaN  Hong Kong  1/24/2011   24   
69679  1/24/2011 14:24       3.25         NaN  Hong Kong  1/24/2011   24   

      month  year  
70758     1  2011  
70759     1  2011  
70760     1  2011  
70761     1  2011  
70762     1  2011  
...     ...   ...  
69675     1  2011  
69676     1  2011  
69677     1  2011  
69678     1  2011  
69679     1  2011  

[158 rows x 12 columns]

In [ ]:
Asia_data_Feb = Asia_data[(Asia_data['month']==2)]

In [ ]:
Asia_data_Feb

InvoiceNo StockCode                          Description  Quantity  \
89570     543911     21485     RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                   DOORMAT UNION FLAG         2   
89572     543911     85150        LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                  WASHROOM METAL SIGN        24   
89574     543911     82551               LAUNDRY 15C METAL SIGN        12   
89575     543911     82583                 HOT BATHS METAL SIGN        12   
89576     543911     22413      METAL SIGN TAKE IT OR LEAVE IT         12   
89577     543911     21166           COOK WITH WINE METAL SIGN         12   
89578     543911     21175          GIN + TONIC DIET METAL SIGN        12   
89579     543911     85152      HAND OVER THE CHOCOLATE   SIGN         12   
89580     543911     21908        CHOCOLATE THIS WAY METAL SIGN        12   
89581     543911     21165        BEWARE OF THE CAT METAL SIGN         12   
89582     543911     20961              STRAWBERRY BATH SPONGE         20   
89583     543911     22195         LARGE HEART MEASURING SPOONS        12   
89584     543911     22200             FRYING PAN PINK POLKADOT         6   
89585     543911     22207                FRYING PAN UNION FLAG         6   
89586     543911     85215       ASSORTED CHEESE FRIDGE MAGNETS        72   
89587     543911     21700          BIG DOUGHNUT FRIDGE MAGNETS        72   
89588     543911     84050      PINK HEART SHAPE EGG FRYING PAN        12   
89589     543911     21231       SWEETHEART CERAMIC TRINKET BOX        24   
89590     543911     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
89591     543911     22111         SCOTTIE DOG HOT WATER BOTTLE         6   
89592     543911     21832                 CHOCOLATE CALCULATOR        12   
89593     543911     20985                     HEART CALCULATOR        12   
89594     543911     22112           CHOCOLATE HOT WATER BOTTLE         6   
89595     543911    47590B          PINK HAPPY BIRTHDAY BUNTING        12   
89596     543911     22117     METAL SIGN HER DINNER IS SERVED         12   
89597     543911     82567            AIRLINE LOUNGE,METAL SIGN        12   
89598     543911     21910                  WAY OUT METAL SIGN         12   
89599     543911     21911                   GARDEN METAL SIGN         12   
100810    544838     22915        ASSORTED BOTTLE TOP  MAGNETS         12   
100811    544838     22363                 GLASS JAR MARMALADE          6   
100812    544838     22362         GLASS JAR PEACOCK BATH SALTS         6   
100813    544838     22361    GLASS JAR DAISY FRESH COTTON WOOL         6   
100814    544838     22553               PLASTERS IN TIN SKULLS        12   
100815    544838     22555            PLASTERS IN TIN STRONGMAN        12   
100816    544838     22556       PLASTERS IN TIN CIRCUS PARADE         12   
100817    544838     20781             GOLD EAR MUFF HEADPHONES         2   
100818    544838     22969         HOMEMADE JAM SCENTED CANDLES        12   
91580     544108     22245         HOOK, 1 HANGER ,MAGIC GARDEN        24   
91581     544108     22244           3 HOOK HANGER MAGIC GARDEN        12   
91582     544108     22243    5 HOOK HANGER RED MAGIC TOADSTOOL        12   
91583     544108     22399   MAGNETS PACK OF 4 CHILDHOOD MEMORY        36   
91584     544108     22629                  SPACEBOY LUNCH BOX         24   
91585     544108     22630                 DOLLY GIRL LUNCH BOX        36   
91586     544108     22326  ROUND SNACK BOXES SET OF4 WOODLAND         48   
91587     544108     22423             REGENCY CAKESTAND 3 TIER        16   
91588     544108     22192                BLUE DINER WALL CLOCK        12   
91589     544108     22193                 RED DINER WALL CLOCK        12   
91590     544108     22433          WATERING CAN GREEN DINOSAUR         6   
91591     544108     22432              WATERING CAN PINK BUNNY         6   
915

In [ ]:
Asia_data_Feb['day'] = pd.to_numeric(Asia_data_Feb['day'], errors='coerce')

<ipython-input-202-1bcd3b82b531>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Asia_data_Feb['day'] = pd.to_numeric(Asia_data_Feb['day'], errors='coerce')


In [ ]:
Asia_data_DF = Asia_data_Feb[Asia_data_Feb['day'] <= 15]

In [ ]:
Asia_data_DF

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   
89575    543911     82583              HOT BATHS METAL SIGN        12   
89576    543911     22413   METAL SIGN TAKE IT OR LEAVE IT         12   
89577    543911     21166        COOK WITH WINE METAL SIGN         12   
89578    543911     21175       GIN + TONIC DIET METAL SIGN        12   
89579    543911     85152   HAND OVER THE CHOCOLATE   SIGN         12   
89580    543911     21908     CHOCOLATE THIS WAY METAL SIGN        12   
89581    543911     21165     BEWARE OF THE CAT METAL SIGN         12   
89582    543911     20961           STRAWBERRY BATH SPONGE         20   
89583    543911     22195      LARGE HEART MEASURING SPOONS        12   
89584    543911     22200          FRYING PAN PINK POLKADOT         6   
89585    543911     22207             FRYING PAN UNION FLAG         6   
89586    543911     85215    ASSORTED CHEESE FRIDGE MAGNETS        72   
89587    543911     21700       BIG DOUGHNUT FRIDGE MAGNETS        72   
89588    543911     84050   PINK HEART SHAPE EGG FRYING PAN        12   
89589    543911     21231    SWEETHEART CERAMIC TRINKET BOX        24   
89590    543911     21232    STRAWBERRY CERAMIC TRINKET BOX        24   
89591    543911     22111      SCOTTIE DOG HOT WATER BOTTLE         6   
89592    543911     21832              CHOCOLATE CALCULATOR        12   
89593    543911     20985                  HEART CALCULATOR        12   
89594    543911     22112        CHOCOLATE HOT WATER BOTTLE         6   
89595    543911    47590B       PINK HAPPY BIRTHDAY BUNTING        12   
89596    543911     22117  METAL SIGN HER DINNER IS SERVED         12   
89597    543911     82567         AIRLINE LOUNGE,METAL SIGN        12   
89598    543911     21910               WAY OUT METAL SIGN         12   
89599    543911     21911                GARDEN METAL SIGN         12   

           InvoiceDate  UnitPrice  CustomerID               Country  \
89570  2/14/2011 12:46       4.95     17829.0  United Arab Emirates   
89571  2/14/2011 12:46       7.95     17829.0  United Arab Emirates   
89572  2/14/2011 12:46       2.55     17829.0  United Arab Emirates   
89573  2/14/2011 12:46       1.45     17829.0  United Arab Emirates   
89574  2/14/2011 12:46       1.45     17829.0  United Arab Emirates   
89575  2/14/2011 12:46       2.10     17829.0  United Arab Emirates   
89576  2/14/2011 12:46       2.95     17829.0  United Arab Emirates   
89577  2/14/2011 12:46       2.08     17829.0  United Arab Emirates   
89578  2/14/2011 12:46       2.55     17829.0  United Arab Emirates   
89579  2/14/2011 12:46       2.10     17829.0  United Arab Emirates   
89580  2/14/2011 12:46       2.10     17829.0  United Arab Emirates   
89581  2/14/2011 12:46       1.69     17829.0  United Arab Emirates   
89582  2/14/2011 12:46       1.25     17829.0  United Arab Emirates   
89583  2/14/2011 12:46       1.65     17829.0  United Arab Emirates   
89584  2/14/2011 12:46       4.25     17829.0  United Arab Emirates   
89585  2/14/2011 12:46       4.25     17829.0  United Arab Emirates   
89586  2/14/2011 12:46       0.65     17829.0  United Arab Emirates   
89587  2/14/2011 12:46       0.85     17829.0  United Arab Emirates   
89588  2/14/2011 12:46       1.65     17829.0  United Arab Emirates   
89589  2/14/2011 12:46       1.25     17829.0  United Arab Emirates   
89590  2/14/2011 12:46       1.25     17829.0  United Arab Emirates   
89591  2/14/2011 12:46       4.95     17829.0  United Arab Emirates   
89592  2/14/2011 12:46       1.65     17829.0  United Arab Emirates   
89593  2/14/2011 12:46       1.25     

### เอาตารางที่หาวันรวมกัน

In [ ]:
New_Asia_data=pd.concat([Asia_data_DJ,Asia_data_DF])

In [ ]:
New_Asia_data

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   
...         ...       ...                               ...       ...   
89595    543911    47590B       PINK HAPPY BIRTHDAY BUNTING        12   
89596    543911     22117  METAL SIGN HER DINNER IS SERVED         12   
89597    543911     82567         AIRLINE LOUNGE,METAL SIGN        12   
89598    543911     21910               WAY OUT METAL SIGN         12   
89599    543911     21911                GARDEN METAL SIGN         12   

           InvoiceDate  UnitPrice  CustomerID               Country  \
70758  1/25/2011 13:26       0.36     12744.0             Singapore   
70759  1/25/2011 13:26       2.95     12744.0             Singapore   
70760  1/25/2011 13:26       2.95     12744.0             Singapore   
70761  1/25/2011 13:26       0.21     12744.0             Singapore   
70762  1/25/2011 13:26       0.21     12744.0             Singapore   
...                ...        ...         ...                   ...   
89595  2/14/2011 12:46       5.45     17829.0  United Arab Emirates   
89596  2/14/2011 12:46       2.95     17829.0  United Arab Emirates   
89597  2/14/2011 12:46       2.10     17829.0  United Arab Emirates   
89598  2/14/2011 12:46       1.65     17829.0  United Arab Emirates   
89599  2/14/2011 12:46       1.65     17829.0  United Arab Emirates   

            date  day month  year  
70758  1/25/2011   25     1  2011  
70759  1/25/2011   25     1  2011  
70760  1/25/2011   25     1  2011  
70761  1/25/2011   25     1  2011  
70762  1/25/2011   25     1  2011  
...          ...  ...   ...   ...  
89595  2/14/2011   14     2  2011  
89596  2/14/2011   14     2  2011  
89597  2/14/2011   14     2  2011  
89598  2/14/2011   14     2  2011  
89599  2/14/2011   14     2  2011  

[188 rows x 12 columns]

### จัดกลุ่มตาม Transaction (Invoice_No)

In [ ]:
transactionsAsia = New_Asia_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
transactionsAsia

InvoiceNo
541975    [22620, 85183B, 85183A, 16236, 21058, 21967, 2...
542102    [21519, 22697, 22699, 22343, 22344, 22345, 223...
542276    [82551, 21165, 82600, 21754, 21756, 82482, 824...
543911    [21485, 48138, 85150, 82552, 82551, 82583, 224...
Name: StockCode, dtype: object

In [ ]:
min_sup = 0.3

frequentItemsets_Asia = list(apriori(transactionsAsia,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03011931133652015, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03011931133652015, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7402101241642789, lift=17.20537191045263), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7000903342366757, lift=17

In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

In [ ]:
finditem('22386','85099B')

{'JUMBO BAG PINK POLKADOT', 'JUMBO BAG RED RETROSPOT'}

มี 2 itemset ที่ผ่าน 0.3

RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03233453810578679,
<br>
<br>
**คนที่ซื้อ 22386(JUMBO BAG PINK POLKADOT) กับ 85099B(JUMBO BAG RED RETROSPOT) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 3.23% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03233453810578679, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22386(JUMBO BAG PINK POLKADOT) แล้วค่อยซื้อ 85099B(JUMBO BAG RED RETROSPOT) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 67.75% มีค่า lift เท่ากับ 8.20 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402),
<br>
<br>
**คนที่ซื้อ 85099B(JUMBO BAG RED RETROSPOT) แล้วค่อยซื้อ 22386(JUMBO BAG PINK POLKADOT)  จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 39.15% มีค่า lift เท่ากับ 8.20 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])

ข้อ 3 เสร็จแล้ว

## 4.หา Frequent Item Set และ Association Rules ของช่วงเวลา Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)

In [ ]:
#เช็คประเทศในทวีปยุโรป
Eu = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [ ]:
#ดึงประเทศในทวีปยุโรป มาจาก data_test
#data_test คือ ตัวที่เคลียร์ข้อมูลเรียบร้อย
Europe_CN = data_test[data_test['Country']==""]
for i in range(len(Eu)):
  x = data_test[data_test['Country']==Eu[i]]
  Europe_CN=pd.concat([x,Europe_CN])

In [ ]:
Europe_CN #เรียกดูตาราง

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
536964  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536965  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536966  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536967  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536968  12/08/2011 10:26      40.00     12358.0         Austria  12/08/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
536964  08    12  2011  
536965  08    12  2011  
536966  08    12  2011  
536967  08    12  2011  
536968  08    12  2011  

[538358 rows x 12 columns]

In [ ]:
#สร้างข้อมูลใหม่ โดยเรียกดูเดือนที่ 12 เพราะเป็นในช่วงเวลาที่ต้องการ
Europe_CN12 = Europe_CN[Europe_CN['month']==12]

In [ ]:
len(set(Europe_CN['Country'])) #ทดไว้ว่าจะดูประเทศของทวีปยุโรป

24

In [ ]:
set(Europe_CN['Country']) #ทดไว้ว่าจะดูประเทศของทวีปยุโรป

{'Austria',
 'Belgium',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Iceland',
 'Italy',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'Spain',
 'Sweden',
 'Switzerland',
 'United Kingdom'}

In [ ]:
Europe_CN12 #เรียกดูตาราง

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
536964  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536965  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536966  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536967  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536968  12/08/2011 10:26      40.00     12358.0         Austria  12/08/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
536964  08    12  2011  
536965  08    12  2011  
536966  08    12  2011  
536967  08    12  2011  
536968  08    12  2011  

[67852 rows x 12 columns]

In [ ]:
#จัดกลุ่มตาม Transaction (Invoice_No)
#แปลไทย จัดกลุ่มตามเลขที่ใบเสร็จการทำธุรกรรม เพื่อดูว่าลูกค้าซื้อสินค้าอะไรบ้าง โดยใช้ตารางEurope_CN12
transactionsEurope_CN12 = Europe_CN12.groupby('InvoiceNo')['StockCode'].apply(list)
transactionsEurope_CN12

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 3027, dtype: object

In [ ]:
#apriori ติดตั้งมาเพื่องานนี้โดยเฉพาะ
#ความถี่ของการทำ item set จริงๆมันคื้อ frequentItemsetsEurope_CN12
#สนใจmin_sup ที่ 0.04 เพราะว่าถ้าเกิน0.04มันไม่มี
min_sup = 0.04
frequentEurope_CN12= list(apriori(transactionsEurope_CN12,min_support = min_sup))
for i in frequentEurope_CN12:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.04790221341261976, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.04790221341261976, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.165023343889778), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.165023343889778)])


In [ ]:
finditem('22086', '22910') #เลือกเพราะว่า มี Confidence เยอะที่สุด

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

มี 1 itemset ที่ผ่าน 0.04

RelationRecord(items=frozenset({'22086', '22910'}), support=0.04790221341261976,
<br>
<br>
**คนที่ซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) กับ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 4.79% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.04790221341261976, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) แล้วค่อยซื้อ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) มีค่า confidence เท่ากับ 51.60% มีค่า lift เท่ากับ 7.17 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.165023343889778),
<br>
<br>
**คนที่ซื้อ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) แล้วค่อยซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) มีค่า confidence ตั้งเท่ากับ 66.51% มีค่า lift เท่ากับ 7.17 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.165023343889778)])

ข้อ 4 เสร็จแล้ว

## 5.หา Frequent Item Set และ Association Rules ของวันวาเลนไทน์

In [ ]:
Valen_data_month = data_test[(data_test['month']==2)]

In [ ]:
Valen_data_Day = Valen_data_month[Valen_data_month['day']=='14']

In [ ]:
Valen_data_Day

InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

           InvoiceDate  UnitPrice  CustomerID         Country       date day  \
88847   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011  14   
88848   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011  14   
88849   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011  14   
88850   2/14/2011 9:24       0.42     17667.0  United Kingdom  2/14/2011  14   
88851   2/14/2011 9:24       4.95     17667.0  United Kingdom  2/14/2011  14   
...                ...        ...         ...             ...        ...  ..   
89950  2/14/2011 15:43       1.25     14606.0  United Kingdom  2/14/2011  14   
89951  2/14/2011 15:43       0.85     14606.0  United Kingdom  2/14/2011  14   
89952  2/14/2011 15:43       1.45     14606.0  United Kingdom  2/14/2011  14   
89953  2/14/2011 15:43       6.95     14606.0  United Kingdom  2/14/2011  14   
89954  2/14/2011 16:39       3.75     14189.0  United Kingdom  2/14/2011  14   

      month  year  
88847     2  2011  
88848     2  2011  
88849     2  2011  
88850     2  2011  
88851     2  2011  
...     ...   ...  
89950     2  2011  
89951     2  2011  
89952     2  2011  
89953     2  2011  
89954     2  2011  

[1108 rows x 12 columns]

### จัดกลุ่มตาม Transaction (Invoice_No)

In [ ]:
transactions_valen = Valen_data_Day.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
len(transactions_valen)

55

In [ ]:
min_sup = 0.09
frequentItemsets_va= list(apriori(transactions_valen,min_support = min_sup))
for i in frequentItemsets_va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22720', '22722'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22722'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.625, lift=5.729166666666667), OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.729166666666667)])
yes
RelationRecord(items=frozenset({'22720', '22969'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22969'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22969'}), confidence=0.625, lift=6.875), OrderedStatistic(items_base=frozenset({'22969'}), items_add=frozenset({'22720'}), confidence=1.0, lift=6.875)])


In [ ]:
set(data[(data['StockCode']=='22720') | (data['StockCode']=='22722') ]['Description'])

{'SET OF 3 CAKE TINS PANTRY DESIGN ', 'SET OF 6 SPICE TINS PANTRY DESIGN', nan}

In [ ]:
finditem('22720','22722')

{'SET OF 3 CAKE TINS PANTRY DESIGN ', 'SET OF 6 SPICE TINS PANTRY DESIGN', nan}

มี 2 itemset ที่ผ่าน 0.09

RelationRecord(items=frozenset({'22720', '22722'}), support=0.09090909090909091,
<br>
<br>
**คนที่ซื้อ 22720(SET OF 3 CAKE TINS PANTRY DESIGN) กับ 22722(SET OF 6 SPICE TINS PANTRY DESIGN) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 9.09% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22722'}), confidence=0.09090909090909091, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22720(SET OF 3 CAKE TINS PANTRY DESIGN) แล้วค่อยซื้อ 22722(SET OF 6 SPICE TINS PANTRY DESIGN) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 62.50% มีค่า lift เท่ากับ 5.729 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ แปรผันตรงกัน

OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.625, lift=5.729166666666667),
<br>
<br>
**คนที่ซื้อ 22722(SET OF 6 SPICE TINS PANTRY DESIGN) แล้วค่อยซื้อ 22720(SET OF 3 CAKE TINS PANTRY DESIGN) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 83.34% มีค่า lift เท่ากับ 5.729 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ แปรผันตรงกัน

OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.729166666666667)])

ข้อ 5 เสร็จแล้ว

## 6.หา Frequent Item Set และ Association Rules โดยจัดกลุ่มตามประเทศ+เดือน

### จัดกลุ่มตามประเทศ+เดือน

In [ ]:
transactions_count_month = data_test.groupby(['Country', 'month'])['StockCode'].apply(list)
transactions_count_month

Country      month
Australia    1        [22449, 22450, 22451, 84970L, 71459, 22429, 22...
             2        [23231, 23230, 23233, 22989, 22990, 22969, 229...
             3        [20665, 22232, 23052, 23053, 20749, 20750, 475...
             4                                                  [22990]
             5        [22457, 23176, 22109, 85014A, 22488, 21524, 23...
                                            ...                        
Unspecified  11       [22326, 20725, 22628, 22631, 84997D, 23254, 23...
             04       [20685, 22691, 48116, 21213, 21977, 21976, 212...
             05       [22953, 23078, 22619, 21888, 21890, 22620, 218...
             07       [21634, 21212, 22964, 23254, 23256, 84378, 209...
             09       [22904, 21329, 21992, 20772, 23196, 22543, 225...
Name: StockCode, Length: 419, dtype: object

In [ ]:
min_sup = 0.27
frequentItemsets_count_month = list(apriori(transactions_count_month,min_support = min_sup))
for i in frequentItemsets_count_month:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'POST', '22326'}), support=0.29594272076372313, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '22326'}), confidence=0.29594272076372313, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8435374149659863, lift=1.497636342672662), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5254237288135593, lift=1.497636342672662)])
yes
RelationRecord(items=frozenset({'22423', 'POST'}), support=0.27923627684964203, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423', 'POST'}), confidence=0.27923627684964203, lift=1.0), OrderedStatistic(items_base=frozenset({'22423'}), items_add=frozenset({'POST'}), confidence=0.6842105263157896, lift=1.214763603925067), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22423'}), confidence=0.4957627118644068, lift=1.2147636039250

In [ ]:
set(data[(data['StockCode']=='POST') | (data['StockCode']=='22326') ]['Description'])

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

In [ ]:
finditem('POST','22326')

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

มี 2 itemset ที่ผ่าน 0.27

RelationRecord(items=frozenset({'POST', '22326'}), support=0.29594272076372313,
<br>
<br>
**คนที่ซื้อ POST(POSTAGE) กับ 22326(ROUND SNACK BOXES SET OF4 WOODLAND) จาก dataset มีค่า confidance ทั้งหมดเท่ากับ 29.59% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '22326'}), confidence=0.29594272076372313, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22326(ROUND SNACK BOXES SET OF4 WOODLAND)แล้วค่อยซื้อ POST(POSTAGE) จาก dataset มีค่า confidance ทั้งหมดเท่ากับ 84.35% มีค่า lift เท่ากับ 1.498 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8435374149659863, lift=1.497636342672662),
<br>
<br>
**คนที่ซื้อ POST(POSTAGE)แล้วค่อยซื้อ 22326(ROUND SNACK BOXES SET OF4 WOODLAND)  จาก dataset มีค่า confidance ทั้งหมดเท่ากับ 52.54% มีค่า lift เท่ากับ 1.498 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5254237288135593, lift=1.497636342672662)])

ข้อ 6 เสร็จแล้ว